Cps/implement sdof statik solver

BasselSaridar opened the pull request:
This Branch is a clean up version of cps/implement_rigid_2D_wing_potential_flow_fsi_example_naca0012 in which the commits are subdivided into 4 Different Steps. The reason why they are all combined in one pull request is that they are all complementary to each other. The steps are the following:

Step_1: All modifications needed for the CompressiblePotentialFlowApplication are done in this first step.
Step_2: The rhombus static sdof test is added into EmpireApplication/tests/sdof_static_solver and the naca0012 example is added to EmpireApplication/test_examples/CoSim_DevExamples/potential_flow_rigid_2D_wing_fsi 
Step_3: All Files created inside the EmpireApplication are added and corresponding modifications to use these files are made here
Step_4: This last commit was made to make the solver run inside potential_flow_solver. The reason is that whenever the simulation was getting into a new wake element in which the nodes have only ONE DoF instead of TWO (VELOCITY_POTENTIAL + AUXILIARY_VELOCITY_POTENTIAL), it is causing Segmentation Fault leading to Code Dump. 

If you have any other suggestion on how to fix it in another way or even any modifications needed elsewhere, please share your ideas here.

Thank You

BasselSaridar committed:
StaticSDoFSolver FSI solver using CompressiblePotentialFlowApplication
Step_1:Modifying/Adding files in CompressiblePotentialFlowApplication

BasselSaridar committed:
StaticSDoFSolver FSI solver using CompressiblePotentialFlowApplication
Step_2:Adding test files in EmpireApplication to run them later when finishing all commits

BasselSaridar committed:
StaticSDoFSolver FSI solver using CompressiblePotentialFlowApplication
Step_3:Modifying/Adding files in EmpireApplication

BasselSaridar committed:
StaticSDoFSolver FSI solver using CompressiblePotentialFlowApplication
Step_4:Making the Solver Initialize Matrices in order to add corresponding DoFs for NEW Wake Elements

BasselSaridar committed:
Changes needed from automatic review

BasselSaridar commented:
@inigolcanalejo @LHoeffner

inigolcanalejo commented:
You can check Codacy

inigolcanalejo reviewed (changes requested):
Good job with the cleaning! Even though the changes are complementary to each other (although the sdof for instance could be added separately) I would still split the PR in smaller ones to make the review faster.

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/custom_conditions/potential_wall_condition.cpp:
```diff
@@ -174,7 +174,7 @@ void PotentialWallCondition<TDim, TNumNodes>::GetDofList(DofsVectorType& Conditi
 
 template <unsigned int TDim, unsigned int TNumNodes>
-void PotentialWallCondition<TDim, TNumNodes>::FinalizeSolutionStep(ProcessInfo& rCurrentProcessInfo)
+void PotentialWallCondition<TDim, TNumNodes>::FinalizeNonLinearIteration(ProcessInfo& rCurrentProcessInfo)
```
@marcnunezc is this change ok with you or is it breaking anything in your workflow?

marcnunezc replied:
Hi, it doesnt break anything in my workflow! Why is FinalizeSolutionStep not enough?

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/ale_potential_flow_solver.py:
```diff
@@ -0,0 +1,33 @@
+
+        return is_converged
+
+    # def FinalizeSolutionStep(self):
```
Why are these lines commented? If they are not used you can directly remove them.

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/compute_forces_on_nodes_process.py:
```diff
@@ -29,24 +29,24 @@ def ExecuteFinalizeSolutionStep(self):
         free_stream_velocity = self.body_model_part.ProcessInfo.GetValue(CPFApp.FREE_STREAM_VELOCITY)
-        #TODO: Read density from ProcessInfo once available
-        free_stream_density = 1.225
+        free_stream_density = self.body_model_part.ProcessInfo.GetValue(CPFApp.FREE_STREAM_DENSITY)
```
👍

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/compute_forces_on_nodes_process.py:
```diff
@@ -29,24 +29,24 @@ def ExecuteFinalizeSolutionStep(self):
         KratosMultiphysics.Logger.PrintInfo('ComputeForcesOnNodesProcess', 'Computing reactions on nodes')
 
-        KratosMultiphysics.VariableUtils().SetNonHistoricalVariableToZero(KratosMultiphysics.REACTION, self.body_model_part.Nodes)
+        KratosMultiphysics.VariableUtils().SetHistoricalVariableToZero(KratosMultiphysics.REACTION, self.body_model_part.Nodes)
```
Now that everything is working, did you check whether we really need to have the reaction as a historical variable?

BasselSaridar replied:
If it is not set as HistoricalVariable then the solution would not converge, we will get even 0 forces in all directions

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -32,7 +32,13 @@ def __init__(self, Model, settings):
+        else: self.trailing_edge_model_part = self.fluid_model_part.GetSubModelPart("trailing_edge_model_part")
+
+        if not self.fluid_model_part.HasSubModelPart("wake__elements"):
+            self.wake_elements_sub_modelpart = self.fluid_model_part.CreateSubModelPart("wake__elements")
```
I would call this model part wake_modelpart
```suggestion
            self.wake_modelpart = self.fluid_model_part.CreateSubModelPart("wake__modelpart")
```

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -46,7 +52,10 @@ def __init__(self, Model, settings):
                 node.Set(KratosMultiphysics.SOLID)
 
-    def ExecuteInitialize(self):
+    def SolveSolutionStep(self):
```
The function solvesolutionstep does not exist in the base class. I understand you want to make the naming consisten because you are calling this function within the SolveSolutionStep of your solver, however:

-  Please use one of the functions in the base class: https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python_scripts/python_process.py
- Please check that this change doesnt break the standard potential solver (have you check that the tests in the CompressiblePotentialFlowApplication?)

BasselSaridar replied:
I set it as ExecuteBeforeSolutionLoop(self)
do you think it is fine?

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -46,7 +52,10 @@ def __init__(self, Model, settings):
+    def SolveSolutionStep(self):
+        self.FindWakeElements()
+
+    def FindWakeElements(self):
```
```suggestion
    def _FindWakeElements(self):
```

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -104,9 +113,9 @@ def __MarkWakeElements(self):
-                        KratosMultiphysics.ELEMENTAL_DISTANCES, distances_to_wake)
+                    elem.SetValue(KratosMultiphysics.ELEMENTAL_DISTANCES, distances_to_wake)
                     counter=0
+                    self.wake_elements_sub_modelpart.Elements.append(elem)
```
Use the AddElement function

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -228,4 +237,20 @@ def __MarkWakeTEElement(self):
\ No newline at end of file
+                    elem.SetValue(CPFApp.WAKE, False)
+
+    def CleanMarking(self):
```
👍 
```suggestion
    def _CleanMarking(self):
```

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -228,4 +237,20 @@ def __MarkWakeTEElement(self):
+
+        for elem in self.wake_elements_sub_modelpart.Elements:
+            elem.SetValue(CPFApp.WAKE, False)
+            elem.Reset(KratosMultiphysics.STRUCTURE)
```
The reset of the STRUCTURE can be done in the trailing edge model part

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -110,6 +110,7 @@ def AddVariables(self):
         # Kratos variables
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.FLAG_VARIABLE) # Required for variational_distance_process
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.DISTANCE)
+        self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.REACTION)
```
Do we need it as historical variable?

BasselSaridar replied:
otherwise the REACTION variable will not be recognized by the solver hence aborting.
in model_part.h, I did not find any function that will just initialize the variable.

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -136,3 +139,20 @@ def Initialize(self):
+
+    def SolveSolutionStep(self):
+        self.solver.Clear()
+        self.solver.InitializeSolutionStep()
```
@philbucher is it ok that we do this here? When moving the wake new dofs appear that need to be added to the dof list which was the reason why we were getting segmentation fault. This means we need to create the dof list in each coupling iteration, which is not done typically.

philbucher reviewed (changes requested):
Blocking to discuss the changes in person next week

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/tests/naca0012_small_test/naca0012_small_parameters.json:
```diff
@@ -0,0 +1,157 @@
+{
+    "problem_data"                     : {
+        "problem_name"    : "naca0012",
```
I believe this file was deleted in a previous PR.  Please remove it. The whole naca0012_small_test is tested in the primal solution of the adjoint test

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_predictors/linear_derivative_based_static_predictor.py:
```diff
@@ -0,0 +1,15 @@
+def Create(predictor_settings, solvers, level):
+    return LinearDerivativeBasedStaticPredictor(predictor_settings, solvers, level)
+
+class LinearDerivativeBasedStaticPredictor(CosimulationBasePredictor):
```
As we discussed, you don't need this class since you can just remove the predictor from the json parameters. Therefore, you can just delete this file since you do not need it.

BasselSaridar replied:
That is true, I removed it

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/kratos_potential_flow_solver.py:
```diff
@@ -0,0 +1,66 @@
+        self.wake_process.CleanMarking()
+
+        self.wake_process.SolveSolutionStep()
+        #super(KratosPotentialFlowSolver, self).InitializeSolutionStep()
```
If you are not using this commented line simply remove it.

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/kratos_potential_flow_solver.py:
```diff
@@ -0,0 +1,66 @@
+
+class KratosPotentialFlowSolver(KratosBaseFieldSolver):
+    def _CreateAnalysisStage(self):
+        #print('debug: project_param. from KratosPotentialFlowSolver:\n',  self.project_parameters)
```
Please clean unused commented lines

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/kratos_potential_flow_solver.py:
```diff
@@ -0,0 +1,66 @@
+
+        self.conversion_process.ExecuteFinalizeSolutionStep()
+        self.lift_process.ExecuteFinalizeSolutionStep()
+        # self.wake_process.CleanMarking()
```
remove unused commented lines

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/kratos_potential_flow_solver.py:
```diff
@@ -0,0 +1,66 @@
+
+    def FinalizeSolutionStep(self):
+        super(KratosPotentialFlowSolver, self).FinalizeSolutionStep()
+        # self.wake_process.CleanMarking()
```
remove unused commented lines

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/sdof_static_solver.py:
```diff
@@ -0,0 +1,126 @@
+        if os.path.isfile(self.output_file_name):
+            os.remove(self.output_file_name)
+    
+    # def OutputSolutionStep(self):
```
Again, ´please remove unused commented lines

BasselSaridar replied:
I uncommented lines 66 and 67 since they were also found in the static_solver
probably this statement is kept commented for debugging when needed

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/sdof_static_solver.py:
```diff
@@ -0,0 +1,126 @@
+        self.x = self.dx + self.x_old
+        self.x_old = self.x
+
+        print('Force Imported = ', self.force)
```
Use rather the logger rather than prints

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/sdof_static_solver.py:
```diff
@@ -0,0 +1,126 @@
+        self.x_old = 0.0
+        self.b = 0.0
+
+        #x contain: displacement and dx is not used here
```
Is this comment updated? As far as I see in the function GetSolutionStepValue you are retrieving dx and not x.

BasselSaridar replied:
Yes all unnecessary variables have been deleted now
self.b, self.x, self.x_old

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/sdof_static_solver.py:
```diff
@@ -0,0 +1,126 @@
+        pass
+
+    def SolveSolutionStep(self):
+        self.b = self.force + self.stiffness*self.x_old
```
Do you actually use self.b?

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/sdof_static_solver.py:
```diff
@@ -0,0 +1,126 @@
+        self.b = self.force + self.stiffness*self.x_old
+        self.dx = self.force/self.stiffness
+
+        self.x = self.dx + self.x_old
```
Do you actually use self.x and self.x_old?

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/simple_steady_coupling_solver.py:
```diff
@@ -0,0 +1,104 @@
+def CreateSolver(cosim_solver_settings, level):
+    return SimpleSteadyCouplingSolver(cosim_solver_settings, level)
+
+class SimpleSteadyCouplingSolver(CoSimulationBaseCouplingSolver):
```
Do we really need this class? As far as I see it is exactly the same as gauss_seidel_strong_coupling_solver.py with the only difference that you override the _SynchronizeInputData/OuputData because we don't have time. Maybe you can derive this class from gauss_seidel_strong_coupling_solver and just override _SynchronizeInputData/OutputData to avoid code duplication?

BasselSaridar replied:
I definitely agree, it is changed now

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_steady_analysis.py:
```diff
@@ -0,0 +1,19 @@
+class CoSimulationSteadyAnalysis(CoSimulationAnalysis):
+
+    def RunSolutionLoop(self):
+        print("")
```
This print can be deleted

inigolcanalejo commented on applications/EmpireApplication/test_examples/CoSim_DevExamples/potential_flow_rigid_2D_wing_fsi/MainCoSim.py:
```diff
@@ -0,0 +1,15 @@
+from __future__ import print_function, absolute_import, division  # makes KratosMultiphysics backward compatible with python 2.6 and 2.7
```
The test example potential_flow_rigid_2D_wing_fsi is way too large. We are trying to keep the size of the repository small so save this test locally by you and only commit the small rhombus test.

inigolcanalejo commented on applications/EmpireApplication/tests/sdof_static_solver/naca0012_small_parameters_coupling.json:
```diff
@@ -0,0 +1,187 @@
+{
+    "problem_data"                     : {
```
As far as I see in this test you are testing not only the static sdof solver but the whole FSI workflow with the potential solver. Therefor I would suggest renaming this test to: fsi_potential_flow_sdof and adding another test where you only test the static sdof solver as it is done as it is done in the sdof_solver (i.e. check sdof_solver)

BasselSaridar replied:
done

inigolcanalejo commented on applications/EmpireApplication/tests/sdof_static_solver/naca0012_test_probe1_ref.dat:
```diff
@@ -0,0 +1,3 @@
+# Results for "element" with Id # 21 at position: x: 1; y: 0.5; z: 0
+# time VELOCITY_POTENTIAL
```
we are now testing the velocity potential using a json file. You can check it in the tests of the compressible potential flow application

inigolcanalejo commented on applications/EmpireApplication/tests/test_EmpireApplication.py:
```diff
@@ -7,6 +7,7 @@
 
 from co_simulation_solver_test_factory import TestKratosSolver
 from co_simulation_solver_test_factory import TestSDoFSolver
+from co_simulation_solver_test_factory import TestSDoFStaticSolver
```
You import the test but not use it, therefore the test will not be run. Please add the test. Check codacy

BasselSaridar replied:
done

inigolcanalejo commented on applications/EmpireApplication/tests/sdof_static_solver/cl_ref.dat:
```diff
@@ -0,0 +1 @@
+ 0.327805503865
```
Instead of using a file to check one value you can check it directly in the test. Check the compressible potential flow application

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
Deleting unnecessary changes

BasselSaridar committed:
All necessary changes due to comments from @InigoLopez

BasselSaridar committed:
Changes according to Codacy

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
Trial for Step Output

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

marcnunezc commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -117,6 +118,12 @@ def AddVariables(self):
     def AddDofs(self):
         KratosMultiphysics.VariableUtils().AddDof(KCPFApp.VELOCITY_POTENTIAL, self.main_model_part)
         KratosMultiphysics.VariableUtils().AddDof(KCPFApp.AUXILIARY_VELOCITY_POTENTIAL, self.main_model_part)
+        KratosMultiphysics.VariableUtils().AddDof(KratosMultiphysics.PRESSURE, self.main_model_part)
```
Do you need to define this Variables in the potential_flow_solver.py? In which cases are you using them?

Maybe you can define them in your ale_potential_flow_solver.py instead?

inigolcanalejo replied:
Hi @marcnunezc the reason why @BasselSaridar added this variables is because the embedded test is currently failing when ran in FullDebug becuase of https://github.com/KratosMultiphysics/Kratos/issues/4834. I just talked with @jcotela and he told me that this will be fixed as soon as https://github.com/KratosMultiphysics/Kratos/pull/5066 is merged.

@BasselSaridar as soon as https://github.com/KratosMultiphysics/Kratos/pull/5066 is merged you can merge master and remove the PRESSURE and VELOCITY_X...

inigolcanalejo replied:
@BasselSaridar now you are able to remove pressure and velocities from the dof list.

BasselSaridar committed:
modifying for test

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
VTK Output is available now

BasselSaridar committed:
Removing the naca0012 examples from test_examples

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
Changes made after June12 without Time Outputs

BasselSaridar committed:
ommit + Push problems

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
Moving the clearing function for re-creating system of DoFs and Matrices
from potetntial_flow_solver ---> to kratos_potetntial_flow_solver

marcnunezc commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -32,9 +32,16 @@ def __init__(self, Model, settings):
+        else: self.trailing_edge_model_part = self.fluid_model_part.GetSubModelPart("trailing_edge_model_part")
+
+        if not self.fluid_model_part.HasSubModelPart("wake__elements"):
+            self.wake_modelpart = self.fluid_model_part.CreateSubModelPart("wake__elements")
```
I would rename this to:

```suggestion
            self.wake_sub_model_part = self.fluid_model_part.CreateSubModelPart("wake__elements")
```

BasselSaridar replied:
in the fsi coupling loop we need to calculate the forces applied, on nodes, in order to transfer them to the structure. If we use the FinializeSolutionStep function then it won't do the job during the coupling and we would get Zero forces on the structure
Using FinalizeNonLinearIteration, we could use the values now in the SolveSolutionStep function inside kratos_potential_flow_solver

marcnunezc commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -32,9 +32,16 @@ def __init__(self, Model, settings):
+            self.trailing_edge_model_part = self.fluid_model_part.CreateSubModelPart("trailing_edge_model_part")
+        else: self.trailing_edge_model_part = self.fluid_model_part.GetSubModelPart("trailing_edge_model_part")
+
+        if not self.fluid_model_part.HasSubModelPart("wake__elements"):
```
Here "wake__elements" has two underscores instead of 1
```suggestion
        if not self.fluid_model_part.HasSubModelPart("wake_elements"):
```

marcnunezc commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -167,3 +174,18 @@ def Initialize(self):
+            msg  = "Fluid solver did not converge for step " + str(self.main_model_part.ProcessInfo[KratosMultiphysics.STEP]) + "\n"
+            msg += "corresponding to time " + str(self.main_model_part.ProcessInfo[KratosMultiphysics.TIME]) + "\n"
+            KratosMultiphysics.Logger.PrintWarning("PotentialFlowSolver",msg)
+            return is_converged
```
For these functions, I would try to use the ones from the FluidSolver as much as possible (the FluidSolver is the PotentialSolver parent). 
Which error do you get if you use the parents method (i.e not defining Predict() FinalizeSolut..() SolveSol..() here)?

philbucher replied:
the fluid solver also check if the time buffer is initialized

Thinking abt it I think setting the `min_buffer_size` to 1 should have the same effect and should be sufficient for potential flow right @inigolcanalejo @marcnunezc ?
=> then you don't have to override those functions

inigolcanalejo replied:
Yes, that should work.

rubenzorrilla replied:
Actually I think that we changed this when we made the PotentialFlowApplication to depend on the FluidDynamicsApplication.

marcnunezc commented on applications/EmpireApplication/tests/sdof_static_solver/results_sdof_static_ref.dat:
```diff
@@ -0,0 +1 @@
+0.002533029591058445
```
If you want to check doubles, you can do that at the python level ([see CompFlow tests](https://github.com/KratosMultiphysics/Kratos/blob/230c3a9ec26e82cc2c5674aca4db53a89594cdb4/applications/CompressiblePotentialFlowApplication/tests/potential_flow_test_factory.py#L63))

philbucher replied:
@BasselSaridar this is not solved is it?

inigolcanalejo replied:
You could save this value in the variable DISPLACEMENT in your sdof solver and then read it in the test.

marcnunezc commented on applications/EmpireApplication/tests/fsi_potential_flow_sdof/naca0012_mesh_movement_Y_ref.dat:
```diff
@@ -0,0 +1,3 @@
+# Results for "node" with Id # 9 at position: x: 0.49032; y: -0.00891; z: 0
+# time MESH_DISPLACEMENT_Y
+0.0 5.086125656643907e-06
```
If you want to check doubles, you can do that at the python level ([see CompFlow tests](https://github.com/KratosMultiphysics/Kratos/blob/230c3a9ec26e82cc2c5674aca4db53a89594cdb4/applications/CompressiblePotentialFlowApplication/tests/potential_flow_test_factory.py#L63))

inigolcanalejo commented:
@marcnunezc regarding https://github.com/KratosMultiphysics/Kratos/pull/5022#discussion_r294167408 in the FSI workflow FinalizeSolutionStep is called only once at the end but we actually need to compute the pressure on the condition in each FSI (nonlinear) iteration.

BasselSaridar committed:
change of wake_sub_model_part name

BasselSaridar committed:
Correct Change

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
Minor Clean Up

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
removing unnecessary DoFs for running CompressiblePotentialFlowApp test

BasselSaridar committed:
Running test with genral results checked - Need to implemenet Displacement Check

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

philbucher commented:
@BasselSaridar @inigolcanalejo @marcnunezc overall to me this looks ready to be merged
@BasselSaridar there are some minor comments that you haven't considered yet but otherwise I would give :+1:

BasselSaridar committed:
Adding Test comparing LIFT_COEFFICIENT and MESH_DISPLACEMENT_Y
LIFT_COEFFICIENT had to be added in the AddVariables function in potential_flow_solver

BasselSaridar committed:
setting define_wake_2d back to how it was

philbucher commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -126,6 +126,7 @@ def AddVariables(self):
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NODAL_H) # Required for modify_distance_process
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.VELOCITY) # Required for modify_distance_process
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.PRESSURE) # Required for modify_distance_process
+        self.main_model_part.AddNodalSolutionStepVariable(KCPFApp.LIFT_COEFFICIENT)
```
why does this have to be historical?

inigolcanalejo replied:
It doesnt need to be historical, @BasselSaridar could you remove it?

BasselSaridar committed:
merge fix from master/needs to be changed later

BasselSaridar committed:
NonHistorical LIFT_COEFFICIENT for check in test

inigolcanalejo commented:
@BasselSaridar you can check codacy, there are some unused imports

BasselSaridar committed:
Codacy Errors solved

BasselSaridar committed:
New Way to addVariables

BasselSaridar committed:
Removing unused Variables

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
adapting code to new format

BasselSaridar committed:
formatting

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/ale_potential_flow_solver.py:
```diff
@@ -0,0 +1,28 @@
+        for mesh_solver in self.mesh_motion_solvers:
+            is_converged &= mesh_solver.SolveSolutionStep()
+
+        is_converged = self.fluid_solver.SolveSolutionStep()
```
In here you are overwriting the mesh solver `is_converged`. I'd do something like `return mesh_is_converged & fluid_is_converged`.

rubenzorrilla commented on applications/CompressiblePotentialFlowApplication/python_scripts/define_wake_process_2d.py:
```diff
@@ -51,9 +51,16 @@ def ExecuteInitialize(self):
 
```
Is this file intentionally changed in this PR?

BasselSaridar replied:
Trailing_edge and wake_elements model parts and lists had to be added in order to make the fsi simulation run under __init__ 
appending the elements inside the lists had to be added as well inside __MarkWakeElements
and the cleaning part was added long time ago. I think it is better to leave it in case someone needs to check its behavior in the python code, even though it is not called in the coupled solver anymore

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
seperating convergence criteria for mesh and fluid solver

BasselSaridar committed:
removing unnecessary variables

BasselSaridar committed:
Fixing is_convergence

inigolcanalejo commented:
@BasselSaridar check codacy, there is still an unused import

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -168,3 +170,15 @@ def Initialize(self):
     def AdvanceInTime(self, current_time):
         raise Exception("AdvanceInTime is not implemented. Potential Flow simulations are steady state.")
+
+    def FinalizeSolutionStep(self):
```
@BasselSaridar lets discuss this next week

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -168,3 +170,15 @@ def Initialize(self):
+    def FinalizeSolutionStep(self):
+        (self.solver).FinalizeSolutionStep()
+
+    def SolveSolutionStep(self):
```
and also this

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -168,3 +170,15 @@ def Initialize(self):
+    def FinalizeSolutionStep(self):
+        (self.solver).FinalizeSolutionStep()
+
+    def SolveSolutionStep(self):
```
I think it should be possible not to reimplement SolveSolutionStep and FinalizeSolutionStep. You might need to define self.min_buffer_size in your solver

BasselSaridar replied:
I set the buffer_size inside ale_potential_flow file as 1 and erased FinalizeSolutionStep() and SolveSolutionStep() from potential_flow_solver and it works perfectly fine.

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
Merge branch 'master' into cps/implement_sdof_statik_solver

BasselSaridar committed:
Changes Requested

BasselSaridar committed:
master merging fix

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -165,6 +167,3 @@ def Initialize(self):
         (self.solver).SetEchoLevel(self.settings["echo_level"].GetInt())
         self.solver.Initialize()
-
-    def AdvanceInTime(self, current_time):
```
Do you need to remove this to make your workflow work?

BasselSaridar replied:
It wasn't used since we do not call for AdvanceinTime() in CoSimulationSteadyAnalysis

inigolcanalejo commented on applications/CompressiblePotentialFlowApplication/python_scripts/potential_flow_solver.py:
```diff
@@ -133,6 +133,8 @@ def AddDofs(self):
         KratosMultiphysics.VariableUtils().AddDof(KCPFApp.VELOCITY_POTENTIAL, self.main_model_part)
         KratosMultiphysics.VariableUtils().AddDof(KCPFApp.AUXILIARY_VELOCITY_POTENTIAL, self.main_model_part)
 
+        KratosMultiphysics.Logger.PrintInfo("PotentialFlowSolver", "Fluid solver DOFs added correctly.")
```
I would add this in another PR and commit only the changes that are necessary for the coupling

BasselSaridar replied:
It is just a minor change, is it worth doing it separately?

inigolcanalejo commented on applications/EmpireApplication/python_scripts/co_simulation_solvers/co_simulation_base_coupling_solver.py:
```diff
@@ -161,3 +161,6 @@ def Check(self):
     def IsDistributed(self):
         return True
+
+    def _Name(self):
```
Same, do you need this in your workflow?

BasselSaridar replied:
removed, and works fine

inigolcanalejo commented on applications/EmpireApplication/tests/fsi_potential_flow_sdof/small_lift_jump_parameters.json:
```diff
@@ -0,0 +1,71 @@
+{
```
why is this file called small_lift_jump? Are you actually using it?

inigolcanalejo commented on applications/EmpireApplication/tests/fsi_potential_flow_sdof/small_lift_jump_parameters.json:
```diff
@@ -0,0 +1,71 @@
+                "epsilon"         : 1e-9
+            }
+        },{
+            "python_module" : "compute_lift_jump_process_2d",
```
This process does not exist any more, I have the feeling you are not using this parameters

inigolcanalejo commented on applications/EmpireApplication/tests/fsi_potential_flow_sdof/small_lift_jump_parameters.json:
```diff
@@ -0,0 +1,71 @@
+            }
+        }],
+        "auxiliar_process_list" :[{
+            "python_module"   : "compare_two_files_check_process",
```
I think this process you are not using any more.

BasselSaridar replied:
The entire process has been removed

inigolcanalejo reviewed (changes requested):
There are some minor changes and some files I think you are not using, but apart from that it is ready for merging. Let me know when you finish and I will remove my blocker.

BasselSaridar committed:
removing small_lift_jump_parameteres and _Name function in co_simulation_base_coupling

BasselSaridar committed:
undo all changes in potential_flow_solver

inigolcanalejo commented:
@philbucher this is ready to be merged

philbucher reviewed (approved):
Looks good to me, thx for the effort

BasselSaridar commented:
Thanks a lot everyone for your cooperation
